### ISSUE TITLE
1.5.0 fails, exiting 0, if pidfile directory does not exist

### ISSUE THREAD

**Issue by n1zyy** [varnish/hitch#312]
tl;dr:

* If you invoke hitch with `--pidfile` pointing to a directory which does not exist, hitch silently exits. (A child process is `clone`d and does exit 1, but it's masked by the parent exiting 0.)
* This problem seems to be out-of-the-box behavior in EPEL (using hitch 1.5.0); the systemd unit tries to write to `/run/hitch/hitch.pid` when `/run/hitch` is not created.

### version info

hitch 1.5.0 on CentOS 7:

```
[root@ip-172-16-1-130 hitch]# yum info hitch
Loaded plugins: fastestmirror
Loading mirror speeds from cached hostfile
 * base: mirrors.advancedhosters.com
 * epel: mirror.cogentco.com
 * extras: mirrors.advancedhosters.com
 * updates: mirrors.advancedhosters.com
Installed Packages
Name        : hitch
Arch        : x86_64
Version     : 1.5.0
Release     : 1.el7
Size        : 221 k
Repo        : installed
From repo   : epel
Summary     : Network proxy that terminates TLS/SSL connections
URL         : https://hitch-tls.org/
License     : BSD
Description : hitch is a network proxy that terminates TLS/SSL connections and forwards the
            : unencrypted traffic to some backend. It is designed to handle 10s of thousands
            : of connections efficiently on multicore machines.

```

### Gory details

(strace output substantially trimmed for brevity)

```
[root@ip-172-16-1-130 hitch]# strace -f /usr/sbin/hitch --pidfile=/run/hitch/hitch.pid --config=/etc/hitch/hitch.conf --daemon

....
fstat(5, {st_mode=S_IFREG|0600, st_size=527, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06e7bd4000
read(5, "0\202\2\v\n\1\0\240\202\2\0040\202\2\0\6\t+\6\1\5\5\0070\1\1\4\202\1\3610\202"..., 4096) = 527
close(5)                                = 0
munmap(0x7f06e7bd4000, 4096)            = 0
open("/etc/localtime", O_RDONLY|O_CLOEXEC) = 5
fstat(5, {st_mode=S_IFREG|0644, st_size=118, ...}) = 0
fstat(5, {st_mode=S_IFREG|0644, st_size=118, ...}) = 0
mmap(NULL, 4096, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7f06e7bd4000
read(5, "TZif2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\1\0\0\0\1\0\0\0\0"..., 4096) = 118
lseek(5, -62, SEEK_CUR)                 = 56
read(5, "TZif2\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\1\0\0\0\1\0\0\0\0"..., 4096) = 62
close(5)                                = 0
munmap(0x7f06e7bd4000, 4096)            = 0
geteuid()                               = 0
chdir("/")                              = 0
clone(child_stack=0, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f06e7bc6b10) = 12202
exit_group(0)                           = ?
+++ exited with 0 +++
strace: Process 12202 attached
set_robust_list(0x7f06e7bc6b20, 24)     = 0
open("/dev/null", O_RDONLY)             = 5
dup3(5, 0, 0)                           = 0
close(5)                                = 0
open("/dev/null", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 5
dup3(5, 1, 0)                           = 1
close(5)                                = 0
open("/dev/null", O_WRONLY|O_CREAT|O_TRUNC, 0666) = 5
dup3(5, 2, 0)                           = 2
close(5)                                = 0
setsid()                                = 12202
open("/run/hitch/hitch.pid", O_WRONLY|O_CREAT|O_NONBLOCK, 0644) = -1 ENOENT (No such file or directory)
exit_group(1)                           = ?
+++ exited with 1 +++
```

The obvious solution is for me to not pass in a pidfile in a non-existent directory. However, the stock configuration for the systemd unit specifies `/run/hitch/hitch.pid` which was not created:

```
[root@ip-172-16-1-130 hitch]# systemctl status hitch
● hitch.service - Network proxy that terminates TLS/SSL connections
   Loaded: loaded (/usr/lib/systemd/system/hitch.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Sat 2019-07-20 00:52:16 UTC; 19s ago
  Process: 12262 ExecStart=/usr/sbin/hitch --pidfile=/run/hitch/hitch.pid --config=/etc/hitch/hitch.conf (code=exited, status=0/SUCCESS)
 Main PID: 12262 (code=exited, status=0/SUCCESS)

Jul 20 00:52:05 ip-172-16-1-130.ec2.internal systemd[1]: Started Network proxy that terminates TLS/SSL connections.
Jul 20 00:52:16 ip-172-16-1-130.ec2.internal systemd[1]: Stopping Network proxy that terminates TLS/SSL connections...
Jul 20 00:52:16 ip-172-16-1-130.ec2.internal systemd[1]: Stopped Network proxy that terminates TLS/SSL connections.
[root@ip-172-16-1-130 hitch]# systemctl start hitch
[root@ip-172-16-1-130 hitch]# systemctl status hitch
● hitch.service - Network proxy that terminates TLS/SSL connections
   Loaded: loaded (/usr/lib/systemd/system/hitch.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Sat 2019-07-20 00:52:39 UTC; 3s ago
  Process: 12285 ExecStart=/usr/sbin/hitch --pidfile=/run/hitch/hitch.pid --config=/etc/hitch/hitch.conf (code=exited, status=0/SUCCESS)
 Main PID: 12285 (code=exited, status=0/SUCCESS)
```

Indeed, running it manually (and even enabling debug logging) does not seem to indicate any problem:

```
[root@ip-172-16-1-130 hitch]# /usr/sbin/hitch --pidfile=/run/hitch/hitch.pid --config=/etc/hitch/hitch.conf -L2
[root@ip-172-16-1-130 hitch]# echo $?
0
```

Simply creating it with `mkdir /run/hitch/` seems to cause it to run fine.

**Comment by n1zyy** [comment]
OK, wait, this is nuttier than I thought. At some point when starting, `/run/hitch` is _removed_!

```
[root@ip-172-16-1-130 hitch]# mkdir /run/hitch
[root@ip-172-16-1-130 hitch]# systemctl status hitch
● hitch.service - Network proxy that terminates TLS/SSL connections
   Loaded: loaded (/usr/lib/systemd/system/hitch.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Sat 2019-07-20 01:12:20 UTC; 2min 5s ago
  Process: 12573 ExecStart=/usr/sbin/hitch --pidfile=/run/hitch/hitch.pid --config=/etc/hitch/hitch.conf (code=exited, status=0/SUCCESS)
 Main PID: 12573 (code=exited, status=0/SUCCESS)

Jul 20 01:12:20 ip-172-16-1-130.ec2.internal systemd[1]: Started Network proxy that terminates TLS/SSL connections.
[root@ip-172-16-1-130 hitch]#
[root@ip-172-16-1-130 hitch]# ps aux | grep hitch
root     12584  0.0  0.0 112708   984 pts/0    R+   01:14   0:00 grep --color=auto hitch
[root@ip-172-16-1-130 hitch]#
[root@ip-172-16-1-130 hitch]#
[root@ip-172-16-1-130 hitch]# systemctl start hitch
[root@ip-172-16-1-130 hitch]# systemctl status hitch
● hitch.service - Network proxy that terminates TLS/SSL connections
   Loaded: loaded (/usr/lib/systemd/system/hitch.service; enabled; vendor preset: disabled)
   Active: inactive (dead) since Sat 2019-07-20 01:14:40 UTC; 3s ago
  Process: 12591 ExecStart=/usr/sbin/hitch --pidfile=/run/hitch/hitch.pid --config=/etc/hitch/hitch.conf (code=exited, status=0/SUCCESS)
 Main PID: 12591 (code=exited, status=0/SUCCESS)

Jul 20 01:14:40 ip-172-16-1-130.ec2.internal systemd[1]: Started Network proxy that terminates TLS/SSL connections.
[root@ip-172-16-1-130 hitch]# ps aux | grep hitch
root     12597  0.0  0.0 112708   984 pts/0    R+   01:14   0:00 grep --color=auto hitch
[root@ip-172-16-1-130 hitch]#
[root@ip-172-16-1-130 hitch]# ls /run/hitch
ls: cannot access /run/hitch: No such file or directory
[root@ip-172-16-1-130 hitch]#
```

Going to dig a bit to try to figure out where it's being deleted... It's possible this is actually a systemd issue.

**Comment by n1zyy** [comment]
I am a dummy. /run/hitch is created by `RuntimeDirectory=hitch` and exists only as long as the service is running, so the behavior with it not existing is expected. Here's the file for posterity:

```
[root@ip-172-16-1-130 hitch]# systemctl cat hitch
# /usr/lib/systemd/system/hitch.service
[Unit]
Description=Network proxy that terminates TLS/SSL connections
After=syslog.target network.target

[Service]

#
# If you want to make changes to this file, please copy it to
# /etc/systemd/system/hitch.service and make your changes there.
# This will override the file kept at /lib/systemd/system/hitch.service
#
# Configuration may be done in /etc/hitch/hitch.params
#

PIDFile=/run/hitch/hitch.pid

# Maximum size of the corefile.
LimitCORE=infinity

RuntimeDirectory=hitch
Type=simple
PrivateTmp=true
ExecStart=/usr/sbin/hitch --pidfile=/run/hitch/hitch.pid --config=/etc/hitch/hitch.conf
ExecReload=/usr/bin/kill -HUP $MAINPID

[Install]
WantedBy=multi-user.target
```

I cannot seem to get hitch to start through systemd, though if I change the `ExecStart` command to run it inside strace logging to a file so I can figure out where it's going wrong, it runs fine. 😦 But this is now weirdness for me to poke at, not an apparent hitch bug, so I'm going to close this out as my initial filing was rather inaccurate.
